### ISSUE TITLE
list(x) infers wrong element type when iterable has conflicting Sequence bases from multiple inheritance

### ISSUE THREAD

**Issue by rhuanbarreto** [astral-sh/ty#3748]
## Summary

`list(x)` infers a different (incorrect) element type compared to `iter(x)`, `list(iter(x))`, and `[e for e in x]` when the iterable class has conflicting `Sequence` parameterizations from multiple inheritance.

## Reproduction

```python
from collections import UserList, abc
from typing import Generic, TypeVar
from abc import ABC

class HasExternalId:
    external_id: str

class HasInternalId:
    internal_id: int

T = TypeVar("T")

class ResourceList(UserList, Generic[T]):
    def __iter__(self):  # type: ignore
        return super().__iter__()

class ExternalIDMixin(abc.Sequence[HasExternalId], ABC): ...
class InternalIDMixin(abc.Sequence[HasInternalId], ABC): ...
class IdMixin(ExternalIDMixin, InternalIDMixin): ...

class MyResource:
    name: str

class MyResourceList(ResourceList[MyResource], IdMixin): ...

def get_list() -> MyResourceList:
    return MyResourceList([])

x = get_list()

reveal_type(iter(x))          # Iterator[MyResource]  -- correct
reveal_type(list(iter(x)))    # list[MyResource]      -- correct
reveal_type([e for e in x])   # list[MyResource]      -- correct
reveal_type(list(x))          # list[HasInternalId | Unknown | HasExternalId]  -- WRONG
```

## Expected

All four expressions should infer element type `MyResource`, since `__iter__` correctly resolves to `Iterator[MyResource]`.

## Actual

`list(x)` infers `list[HasInternalId | Unknown | HasExternalId]` while all other iteration patterns correctly infer `MyResource`.

## Analysis

ty correctly resolves `__iter__` to `Iterator[MyResource]` (proven by `iter(x)` and list comprehension). But `list()` appears to use a different resolution path - likely `Sequence.__getitem__` or the `Sequence` element type - that hits the conflicting `Sequence[HasExternalId]` and `Sequence[HasInternalId]` bases from the mixin classes, producing a union instead of using the `__iter__` return type.

## Real-world impact

This affects the [Cognite Python SDK](https://github.com/cognitedata/cognite-sdk-python) (v8.7.0), where `CogniteResourceList` uses this exact pattern (multiple `Sequence` parameterizations via `IdTransformerMixin`). Code like:

```python
files: list[FileMetadata] = list(cdf_client.files.list(...))
```

requires `# ty: ignore[invalid-assignment]` even though the types are correct at runtime and pass in mypy.

## Workarounds

- `list(iter(x))` instead of `list(x)`
- `[e for e in x]` instead of `list(x)`
- `# ty: ignore[invalid-assignment]`

## Environment

- ty: latest (via archgate/astral toolchain)
- Python: 3.14
- OS: Windows 11

**Comment by carljm** [comment]
Thanks for the report!

Just for clarity for anyone else trying to repro: the OP snippet is missing a crucial annotation that is necessary in order to reproduce the mentioned results: `ResourceList.__iter__` needs to be annotated as returning `Iterator[T]`. This matches the original source in Cognite. Without that annotation, all type checkers infer `Iterator[Unknown]` or `list[Unknown]` in all four cases. Here's the corrected snippet:

```py
from collections import UserList, abc
from typing import Generic, TypeVar, Iterator
from abc import ABC

class HasExternalId:
    external_id: str

class HasInternalId:
    internal_id: int

T = TypeVar("T")

class ResourceList(UserList, Generic[T]):
    def __iter__(self) -> Iterator[T]:
        return super().__iter__()

class ExternalIDMixin(abc.Sequence[HasExternalId], ABC): ...
class InternalIDMixin(abc.Sequence[HasInternalId], ABC): ...
class IdMixin(ExternalIDMixin, InternalIDMixin): ...

class MyResource:
    name: str

class MyResourceList(ResourceList[MyResource], IdMixin): ...

def get_list() -> MyResourceList:
    return MyResourceList([])

x = get_list()

reveal_type(iter(x))          # Iterator[MyResource]  -- correct
reveal_type(list(iter(x)))    # list[MyResource]      -- correct
reveal_type([e for e in x])   # list[MyResource]      -- correct
reveal_type(list(x))          # list[HasInternalId | Unknown | HasExternalId]  -- WRONG
```

I think the class is basically invalid here: `IdMixin` is invalid because inheriting both `Sequence[HasExternalId]` and `Sequence[HasInternalId]` is invalid (ty and all other type checkers except for mypy flag this); it's not possible for a single class to be simultaneously a sequence of two distinct types. And I think the definition of `MyResourceList` should also be flagged as a Liskov violation, because its MRO ends up having `__iter__` overridden in a non-substitutable way (only Zuban currently flags this, but I think we should, too.)

Most other type-checkers consider `list(x)` unresolvable, due to this invalid MRO, and end up inferring it as `list[Unknown]` or `list[Any]`. This is true for pyright, mypy, zuban, and pyrefly; only pycroscope infers `list[MyResource]`. But `list[Unknown]` still doesn't cause errors, whereas our union does.

I'm pretty sure that our union here comes from #2799, and should go away at some point as we move to the new solver.

Even then, there will be an interesting question here about whether we end up with `list[Unknown]` or `list[MyResource]`. I think the latter would come from a purely structural match between `MyResource` and `Iterable[T]` (in the constructor signature of `list`), which would hit `MyResource.__iter__` and look no further. Once Liskov errors have been emitted on the definitions of `IdMixin` and `MyResourceList` (which pycroscope does not do), this seems pretty reasonable to me. The extra solutions (`InternalIdMixin` and `ExternalIdMixin`) come from checking for nominal subtyping matches (and finding `Iterable` in the MRO several times, since typeshed says `Sequence` inherits it), which we do before we check for a structural match.

**Comment by carljm** [comment]
The only part of this issue that is a stable blocker is getting the `HasInternalId | HasExternalId` parts out of the final list inference; it should be either `list[Unknown]` or `list[MyResource]`. Most type checkers reveal `list[Unknown]` here, and it's certainly fine for stable release if we do too.
